shared/src/balancer/event_handler.rs

```diff
@@ -237,75 +248,120 @@ impl PoolRegistry {
             .max()
             .unwrap_or(0)
     }
+}
+
+pub struct BalancerPoolRegistry {
+    weighted_pool_updater:
+        Mutex<EventHandler<Web3, BalancerV2WeightedPoolFactoryContract, DynPoolRegistry>>,
+    two_token_pool_updater:
+        Mutex<EventHandler<Web3, BalancerV2WeightedPool2TokensFactoryContract, DynPoolRegistry>>,
+}
 
-    fn contract_to_balancer_events(
+impl BalancerPoolRegistry {
+    pub async fn get_pools_containing_token_pairs(
         &self,
-        contract_events: Vec<EthContractEvent<ContractEvent>>,
-    ) -> Result<Vec<(EventIndex, PoolCreated)>> {
-        contract_events
-            .into_iter()
-            .map(|EthContractEvent { data, meta }| {
-                let meta = match meta {
-                    Some(meta) => meta,
-                    None => return Err(anyhow!("event without metadata")),
-                };
-                match data {
-                    ContractEvent::PoolCreated(event) => convert_pool_created(&event, &meta),
-                }
-            })
-            .collect::<Result<Vec<_>>>()
+        token_pairs: HashSet<TokenPair>,
+    ) -> HashSet<RegisteredWeightedPool> {
```

Comment:
Why a `HashSet` and not a `Vec`? Also, I don't think `Hash` makes sense for this type. For example (in pseudo-code):
```rust
let pool0 = RegisteredWeightedPool {
  pool_id: some_id,
  address: some_address,
  ...
}
let pool1 = RegisteredWeightedPool {
  pool_id: some_id,
  address: some_other_address,
  ...
}
```

What do you expect to happen when adding these pools into the same `HashSet`? Should there be two (which is what will happen now with the `derive(Hash)` implementation)? Or one (because pool IDs are unique)? I would like to avoid this kind of ambiguity.

Additionally, using a `HashSet` here makes the code imply that you expect that there are pools common to both `weighted_pool_updater` and `two_token_pool_updater` which there shouldn't be and makes this quite confusing to me.